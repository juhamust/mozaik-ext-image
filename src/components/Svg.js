import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import FileImageO from 'react-icons/lib/fa/file-image-o'
import { Widget, WidgetHeader, WidgetBody } from '@mozaik/ui'
import Vivus from 'vivus'

const ImageArea = styled.div``

class Svg extends Component {
  componentWillUnmount() {
    if (this.intervalHandle) {
      clearInterval(this.intervalHandle)
    }
  }

  componentDidMount() {
    if (!this._e) {
      return
    }

    const { type, duration } = this.props
    const svgObject =
      this._e.getElementsByTagName('object')[0] || this._e.getElementsByTagName('svg')[0]
    const options = {
      type,
      duration,
    }
    new Vivus(svgObject, options, ref => {
      if (this.props.loop && ref.getStatus() === 'end') {
        ref.reset().play()
      }
    })
  }

  renderInline(content) {
    return {
      __html: content,
    }
  }

  render() {
    const { url } = this.props
    const imageArea = (
      <ImageArea innerRef={e => (this._e = e)}>
        {this.props.content ? (
          <div dangerouslySetInnerHTML={this.renderInline(this.props.content)} />
        ) : (
          <object type="image/svg+xml" data={url} />
        )}
      </ImageArea>
    )

    return (
      <Widget>
        {this.props.title && (
          <WidgetHeader title={this.props.title} subjectPlacement="append" icon={FileImageO} />
        )}
        <WidgetBody style={{ top: this.props.title === undefined && 0 }}>{imageArea}</WidgetBody>
      </Widget>
    )
  }
}

Svg.propTypes = {
  url: PropTypes.string,
  content: PropTypes.string,
  title: PropTypes.string,
  loop: PropTypes.bool,
  duration: PropTypes.number,
  type: PropTypes.oneOf([
    'delayed',
    'sync',
    'oneByOne',
    'script',
    'script',
    'scenario',
    'scenario-sync',
  ]),
}

Svg.defaultProps = {
  loop: false,
  duration: 100,
  type: 'delayed',
}

export default Svg
