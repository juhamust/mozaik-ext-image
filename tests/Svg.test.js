import React from 'react'
import Svg from '../src/components/Svg'

it('matches with snapshot', () => {
  const content = `
  <svg><test /></svg>
  `
  const wrapper = shallow(<Svg content={content} />)
  expect(wrapper).toMatchSnapshot()
})
