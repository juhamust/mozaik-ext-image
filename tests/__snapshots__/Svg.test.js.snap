// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches with snapshot 1`] = `
<Widget
  style={Object {}}
>
  <WidgetBody
    style={
      Object {
        "top": 0,
      }
    }
  >
    <styled.div
      innerRef={[Function]}
    >
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "
            <svg><test /></svg>
            ",
          }
        }
      />
    </styled.div>
  </WidgetBody>
</Widget>
`;
